proc_dash/app.py

```diff
@@ -311,70 +334,118 @@ def toggle_dataset_name_dialog(
     Returns any errors encountered during input file processing as a user-friendly message.
     """
-    if contents is None:
+    # Upload components need to be manually replaced to clear contents,
+    # otherwise previously uploaded imaging/pheno bagels cannot be re-uploaded
+    # (e.g. if a user uploads pheno_bagel.csv, then imaging_bagel.csv, then pheno_bagel.csv again)
+    # see https://github.com/plotly/dash-core-components/issues/816
+    upload_buttons_reset = [
+        dcc.Upload(
+            id={"type": "upload-data", "index": "imaging", "btn_idx": 0},
+            children=dbc.Button(
+                "Drag and Drop or Select an Imaging .csv File",
+                color="secondary",
+            ),
+            multiple=False,
+        ),
+        dcc.Upload(
+            id={"type": "upload-data", "index": "phenotypic", "btn_idx": 1},
+            children=dbc.Button(
+                "Drag and Drop or Select a Phenotypic .csv File",
+                color="secondary",
+            ),
+            multiple=False,
+        ),
+    ]
+
+    if all(c is None for c in contents) and memory_filename is None:
         return (
+            no_update,
             None,
             None,
             "Upload a CSV file to begin.",
             no_update,
             no_update,
             no_update,
+            None,
+            no_update,
         )
 
+    contents = ctx.triggered[0]["value"]
+    filename = filename[ctx.triggered_id.btn_idx]
+
     try:
         bagel, upload_error = util.parse_csv_contents(
-            contents=contents, filename=filename
+            contents=contents,
+            filename=filename,
+            schema=ctx.triggered_id.index,
         )
         if upload_error is None:
-            overview_df = util.get_pipelines_overview(bagel=bagel)
-            pipelines_dict = util.extract_pipelines(bagel=bagel)
+            overview_df = util.get_pipelines_overview(
+                bagel=bagel, schema=ctx.triggered_id.index
+            )
+            pipelines_dict = util.extract_pipelines(
+                bagel=bagel, schema=ctx.triggered_id.index
+            )
     except Exception as exc:
         print(exc)  # for debugging
         upload_error = "Something went wrong while processing this file."
 
     if upload_error is not None:
         return (
+            filename,
             None,
             None,
             f"Error: {upload_error} Please try again.",
             "none",
             None,
             {"display": "none"},
+            None,
+            upload_buttons_reset,
         )
 
     # Change orientation of pipeline dataframe dictionary to enable storage as JSON data
     for key in pipelines_dict:
         pipelines_dict[key] = pipelines_dict[key].to_dict("records")
 
     dataset_summary = util.construct_summary_str(overview_df)
+    column_count_str = f"Total number of columns: {len(overview_df.columns)}"
```

Comment:
> I don't think this number can ever change, no? I.e. it doesn't have to be recomputed.

Not sure I understand what you mean. The number refers to the number of columns in the visible datatable in the dashboard, which will vary depending on the number of unique assessments/pipelines in the uploaded bagel. And therefore, needs to be recomputed whenever a new bagel is uploaded.

Reply:
Ah, no, all good. I just confused myself here. I got the part about the columns of the dataframe, I just mistakenly assumed this would be recomputed everytime a filter condition changes, not just when a new bagel.csv is uploaded!